contracts/strategy/aave/UniversalSwapper.sol

```diff
@@ -0,0 +1,130 @@
+     * is to help set the mapping. It can be called internally during
+     * initialization, through permissioned functions, etc.
+     */
+    function _setUniFees(address _token0, address _token1, uint24 _fee) internal {
+        uniFees[_token0][_token1] = _fee;
+        uniFees[_token1][_token0] = _fee;
+    }
+
+    /**
+     * @dev Used to swap a specific amount of `_from` to `_to`.
+     * This will check and handle all allowances as well as not swapping
+     * unless `_amountIn` is greater than the set `_minAmountOut`.
+     *
+     * If one of the tokens matches with the `base` token it will do only
+     * one jump, otherwise will do two jumps.
+     *
+     * The corresponding uniFees for each token pair will need to be set
+     * otherwise, this function will revert.
+     *
+     * @param _from The token we are swapping from.
+     * @param _to The token we are swapping to.
+     * @param _amountIn The amount of `_from` we will swap.
+     * @param _minAmountOut The min of `_to` to get out.
+     * @return _amountOut The actual amount of `_to` that was swapped to
+     */
+    function _swapFrom(address _from, address _to, uint256 _amountIn, uint256 _minAmountOut) internal returns (uint256 _amountOut) {
+        _checkAllowance(permit2, _from, _amountIn);
+
+        bytes memory path;
+        bytes memory commands;
+        bytes[] memory inputs;
+
+        if (_from == base || _to == base) {
+            path = abi.encodePacked(_from, uniFees[_from][_to], _to);
+        } else {
+            path = abi.encodePacked(
+                _from, // tokenIn
+                uniFees[_from][base], // from-base fee
+                base, // base token
+                uniFees[base][_to], // base-to fee
+                _to // tokenOut
+            );
+        }
+
+        (commands, inputs) = _encodeSwapExactInput(
+            address(this),
+            _amountIn,
+            _minAmountOut,
+            path,
+            true
+        );
+
+        IPermit2(permit2).approve(_from, router, uint160(_amountIn), uint48(block.timestamp));
+        uint256 balanceBefore = ERC20(_to).balanceOf(address(this));
+        IUniversalRouter(router).execute(commands, inputs, block.timestamp);
+        ERC20(_from).safeApprove(permit2, 0);
+
+        _amountOut = ERC20(_to).balanceOf(address(this)) - balanceBefore;
+    }
+
+    /**
+     * @dev Internal safe function to make sure the contract you want to
+     * interact with has enough allowance to pull the desired tokens.
+     *
+     * @param _contract The address of the contract that will move the token.
+     * @param _token The ERC-20 token that will be getting spent.
+     * @param _amount The amount of `_token` to be spent.
+     */
+    function _checkAllowance(address _contract, address _token, uint256 _amount) internal {
+        if (ERC20(_token).allowance(address(this), _contract) < _amount) {
+            ERC20(_token).approve(_contract, 0);
```

Comment:
What is the reason for having two calls with the first 0 approved?

Reply:
I found that some erc20 like USDT have the requirement to approve 0 before real approval. Here https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
I don't see the issue having this here. Better to keep safier. Merging.